0101-nix formatting.md

```diff
@@ -0,0 +1,672 @@
+  , fetchurl
+  , ...
+  } @ attrs:
+  stdenv.mkDerivation # ...
+  ```
+
+- This is not consistent with _Let-In_, and _AttrSet_,
+  where the separator goes after the element
+  instead of at the beginning.
+- It ruins "folding by indentation" modes
+  on Vim, Neovim, VSCode, and other major code editors,
+  because the data-structure has the same indentation
+  as the opening brace.
+
+❌ Bad:
+
+```nix
+{ mkDerivation, lib, fetchurl, ... }@attrs: stdenv.mkDerivation # ...
+```
+
+- One-liners are unreadable and produce bad diffs.
+
+❌ Bad:
+
+```nix
+{ mkDerivation, lib, fetchurl, extra-cmake-modules, kdoctools, wrapGAppsHook
+, karchive, kconfig, kcrash, kguiaddons, kinit, kparts, kwind, ... }@attrs:
+stdenv.mkDerivation # ...
+```
+
+- It's hard to tell this destructuring has an ellipsis (`...`) at a first glance,
+  because it's mixed with the other arguments.
+- Moving elements becomes harder
+  than a simple whole-line movement.
+  (Moving a whole line is normally a keyboard-shortcut
+  or command in major code editors).
+- Excessively compact:
+  adding, removing, or editing an argument
+  produces a diff in more than one argument.
+- `}@attrs` is not intuitive
+  with the rules of written english,
+  where you add whitespace
+  after the end of the previous phrase
+  (`phrase. Other phrase`).
+
+### Inherit
+
+✅ Good:
+```nix
+inherit foo bar;
+inherit (attrs) foo bar;
+inherit
+  foo
+  bar
+  ;
+inherit (attrs)
+  foo
+  bar
+  ;
+```
+
+- Inherit expressions can have their symbols either all on the same line as the `inherit`, or one line per element each.
+- The `(attrs)` is always on the same line as the `inherit` statement. There are no spaces on the inside of its parentheses.
+
+❌ Bad:
+```nix
+TODO
+inherit (attrs) foo
+  bar
+  ;
+```
+
+❌ Bad:
+```nix
+inherit
+  (attrs)
+  foo
+  bar
+  ;
+```
+
+✅ Good:
+
+```nix
+{
+  inherit
+    (mkFruit {
+      a = "apple";
+      b = "banana";
+      c = "cherry";
+      d = "date";
+      e = "elderberry";
+    })
+    feijoa
+    grape
+    honeydew
+    ;
+}
+```
+
+- Technically this is inconsistent because the attribute set is sometimes on the same line as the `inherit` and other times not. However, this is a really niche use case and generally you won't need to write such code anyways so we are willing to sacrifice consistency for a better default.
+
+### If-Then-Else
+
+✅ Good:
+
+```nix
+if predicate then
+  foo
+else
+  bar
+```
+
+- "Pulling up" the `then` keyword is consistent with other similar syntax constructs
+- The bodies are always indented
+
+❌ Bad (old):
+
+```nix
+if predicate
+then foo
+else bar
+```
+
+- TODO explain why bad
+- ~~The keyword at the beginning of the line
+  states clearly the meaning of the content that follows.~~
+- ~~Produces a clean diff when you add more code.
+  For example: adding content to the `else`
+  only produces a diff in the `else`.~~
+
+❌ Bad:
+
+```nix
+if predicate then foo else bar
+```
+
+- One-liners are hard to understand,
```

Comment:
I would argue against this in simple cases. Maybe there could be a limit for line length? If it is passed, the one-liner must be expanded.

```
env = {
  inherit version;
  release = if build.debug then 0 else 1;
  minify = if opts.minify then 1 else 0;
  codename = opts.name;
}
compileFlags =
  if build.debug then
    {
      wall = true;
      werror = true;
      optimize = 0;
    }
  else
    {
       ...
     }
  ;
```


I'm also not sure how to handle the conditional inside the assignment here. The hanging semicolon feels awkward.

Reply:
Hmm but now we're absorbing braces again. Maybe that's ok? The draft right now specifically doesn't want to do this, and the reasoning is sound in my opinion. I get that not absorbing them indents the attributes another level and adds four additional lines in this case, but really, using the braces like you did doesn't make much sense in nix, as they don't delimit blocks and are thus not part of the control sequence.

But I do agree that we could just add the semicolon at the end. It's not that important semantically;

```nix
let
  compileFlags =
    if build.debug then
      {
        wall = true;
        werror = true;
        optimize = 0;
      }
    else
      {
        _ = "...";
      };
  env = {
    inherit version;
    release = if build.debug then 0 else 1;
    minify = if opts.minify then 1 else 0;
    codename = opts.name;
  };
in { }
```

You see that even without the semicolon, it's very clear that a new attribute starts.